python/lsst/daf/butler/formatters/json.py

```diff
@@ -34,90 +34,45 @@
 import json
 from typing import Any
 
-from .file import FileFormatter
+from lsst.resources import ResourcePath
 
+from .typeless import TypelessFormatter
 
-class JsonFormatter(FileFormatter):
-    """Formatter implementation for JSON files."""
 
-    extension = ".json"
+class JsonFormatter(TypelessFormatter):
+    """Read and write JSON files."""
 
-    unsupportedParameters = None
-    """This formatter does not support any parameters (`frozenset`)"""
+    default_extension = ".json"
+    unsupported_parameters = None
+    can_read_from_uri = True
 
-    def _readFile(self, path: str, pytype: type[Any] | None = None) -> Any:
-        """Read a file from the path in JSON format.
+    def read_from_uri(self, uri: ResourcePath, component: str | None = None, expected_size: int = -1) -> Any:
+        # json.load() reads the entire file content into memory
+        # and is no different from json.loads(uri.read()). It does not attempt
+        # to support incremental reading to minimize memory usage.
+        # This means the JSON string always has to be read entirely into
+        # memory regardless of being remote or local.
+        json_bytes = uri.read()
 
-        Parameters
-        ----------
-        path : `str`
-            Path to use to open JSON format file.
-        pytype : `class`, optional
-            Not used by this implementation.
-
-        Returns
-        -------
-        data : `object`
-            Data as Python object read from JSON file.
-        """
-        with open(path, "rb") as fd:
-            data = self._fromBytes(fd.read(), pytype)
-
-        return data
-
-    def _writeFile(self, inMemoryDataset: Any) -> None:
-        """Write the in memory dataset to file on disk.
-
-        Will look for `_asdict()` method to aid JSON serialization, following
-        the approach of the simplejson module.
-
-        Parameters
-        ----------
-        inMemoryDataset : `object`
-            Object to serialize.
-
-        Raises
-        ------
-        Exception
-            The file could not be written.
-        """
-        self.fileDescriptor.location.uri.write(self._toBytes(inMemoryDataset))
-
-    def _fromBytes(self, serializedDataset: bytes, pytype: type[Any] | None = None) -> Any:
-        """Read the bytes object as a python object.
-
-        Parameters
-        ----------
-        serializedDataset : `bytes`
-            Bytes object to unserialize.
-        pytype : `class`, optional
-            Not used by this implementation.
-
-        Returns
-        -------
-        inMemoryDataset : `object`
-            The requested data as a Python object or None if the string could
-            not be read.
-        """
         try:
-            data = json.loads(serializedDataset)
+            data = json.loads(json_bytes)
```

Comment:
The discussion on https://docs.pydantic.dev/latest/concepts/json/#json-parsing implies that 

1. Trying `pytype.model_validate_json()` here might be significantly faster.
2. We might want to consider using `jiter` for any other JSON if it's meant to be so much faster.
3. @TallJimbo might be right that a special PydanticJsonFormatter is better (although we still have to let JsonFormatter support Pydantic because of backwards compatibility).

@ktlim I assume that `jiter` is part of rubinenv 9 because of the pydantic dependency but maybe it should be explicit rather than explicit if I want to go with point 2 above.

Reply:
Oh. The new json parser is part of pydantic itself: https://docs.pydantic.dev/latest/api/pydantic_core/#pydantic_core.from_json

Reply:
Comparison reading 100MB JSON file:
```
In [8]: %timeit json.loads(data)
539 ms ± 6.83 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [9]: %timeit from_json(data)
266 ms ± 5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
So factor of two improvement but I don't know how many JSON files we are going to be writing that are around that size or larger. (YAML reading the same data is a disaster).